lib/NQP/Config.pm

```diff
@@ -1096,18 +1092,34 @@ sub git_checkout {
     my $config  = $self->config;
     my $options = $self->{options};
     my $pwd     = cwd();
-
+    my $pullurl = $self->repo_url( $repo, action => 'pull' );
+    my $pushurl = $self->repo_url( $repo, action => 'push' );
+
+    # Clone / fetch git reference repo
+    if ( $config->{git_reference} ) {
+        my $ref_dir =
+          $self->reference_dir( $self->{repo_maps}{$repo}[1] );
+        if ( !-d $ref_dir ) {
+            my @ref_args = ( 'git', 'clone', '--bare' );
+            push @ref_args, "--depth=$options->{'git-depth'}"
+              if $options->{'git-depth'};
+            push @ref_args, $pullurl, $ref_dir;
+            $self->msg("Cloning reference from $pullurl\n");
+            system_or_die(@ref_args);
+        }
+        else {
+            chdir($ref_dir);
+            system_or_die( 'git', 'fetch' );
+            chdir($pwd);
+        }
+    }
     # get an up-to-date repository
     if ( !-d $dir ) {
-        my ( $pullurl, $pushurl ) = (
-            $self->repo_url( $repo, action => 'pull' ),
-            $self->repo_url( $repo, action => 'push' ),
-        );
         my @args = ( 'git', 'clone' );
-        if ( $config->{reference} ) {
+        if ( $config->{git_reference} ) {
             my $ref_dir =
-              $self->reference_dir( $self->{repo_maps}{$repo}[1], $dir );
-            die "Can't $repo repository directory in $config->{reference}"
+              $self->reference_dir( $self->{repo_maps}{$repo}[1] );
```

Comment:
I never used referencing (otherwise it would be in working state, of course!). Did you consult with somebody who did? There probably were specific use cases and a reason why `reference_dir` was fed with `$dir`. But your implementation looks rather reasonable to me.

Reply:
If I understand it correctly that was a fallback mechanism. That made sense with the previous behavior, where one had to create the reference repo manually and the Configure.pl only used the files that were there. The new behavior is to create and update the reference directory automatically, so a fallback has no use, as it would be created if not found.

Reply:
Also the only "user" I ever knew of is the rakudobrew program, which used this to not redownload the repos on every `build`. For that use case these changes should work out well.